### PR TITLE
Fixed issue #108301 by escaping URI before creating a MarkdownString.

### LINKED ISSUE #108301: Terminal tooltips render markdown if they span a range that includes colons
If the range returned from a TerminalLinkProvider spans a colon, it seems to cause the tooltip to render markdown instead of being a clickable link.

Normal:

Screenshot 2020-10-08 at 12 57 27

But if I increase the returned `length` slightly to include the space/line/col, it does this:

Screenshot 2020-10-08 at 12 57 49

The `TerminaLink` I'm returning looks like this:

```js
{
	length, // This is the only property changing between the screenshots
	startIndex,
	tooltip: "Open file in editor",
});
```

### PR DESCRIPTION
**Pull request by samrg123**
Fixed issue #108301 by escaping URI with 'encodeURI' before creating a MarkdownString in 'TerminalLinkManager.ts::_getLinkHoverString'.

You can test the original bug and fix by creating an extension that registers a TerminalLinkProvider that returns a TerminalLink spanning a space.

The extension activate function I used for testing was:
```
export function activate(context: vscode.ExtensionContext) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.log('Congratulations, your extension "terminallinktest" is now active!');


	vscode.window.activeTerminal?.show();


	vscode.window.registerTerminalLinkProvider({
		provideTerminalLinks: (context, token) => {

			return [{
				startIndex: 0,
				length: 10,
				tooltip: "crash the tooltip"
			}];
		},

		handleTerminalLink: (link: any) => { }
	});

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with registerCommand
	// The commandId parameter must match the command field in package.json
	let disposable = vscode.commands.registerCommand('terminallinktest.testLink', () => {
		// The code you place here will be executed every time your command is executed

		vscode.window.showInformationMessage("Hello from termainLinkTest");

		// const crashPrefix = "package:test_api/src/frontend/expect.dart";
		// const crashMiddle = " 154:30";
		// const crashSuffix = "		fail";
		// const crashString = crashPrefix + crashMiddle + crashSuffix;

		let text = "";

		// Send all ascii chars
		for(var i = 32; i < 300; ++i) {

			let message = i+":";

			// escape quotes and ticks
			if(i === 34 || i === 96) {
				message = "'"+message+"\\"+String.fromCharCode(i)+"'";

			// escape "'"
			} else if(i === 39) {

				message = '"'+message+'\'"';

			// default
			} else {
				message = "'"+message+String.fromCharCode(i)+"'";
			}

			text+= message+" ";
		}

		vscode.window.activeTerminal?.sendText("echo "+text);
	});

	context.subscriptions.push(disposable);
}

```

This PR fixes #108301

### CODE REVIEW

File: src/vs/workbench/contrib/terminal/browser/links/terminalLinkManager.ts [hunk 1]
```diff
@@ -229,6 +229,9 @@ export class TerminalLinkManager extends DisposableStore {
 			}
 		}
 
+		// encode the uri so TerminalLink get rendered as anchor element https://github.com/microsoft/vscode/issues/108301
+		uri = encodeURI(uri);
```
Comment (rheh) [inline src/vs/workbench/contrib/terminal/browser/links/terminalLinkManager.ts:233]: Suggestion: while changing this section how about moving the undefined check out of the template and naming the variable so not to give the impression the function attribute to being mutated.
```
const encodedUri = encodeURI(uri) || 'undefined';
return new MarkdownString(`[${label || nls.localize('followLink', "Follow Link")}](${encodedUri}) (${clickLabel})`, true);
	
```

**Review by rheh — APPROVED** [review]
One suggestion - other than that code looks good.

File: src/vs/workbench/contrib/terminal/browser/links/terminalLinkManager.ts (hunk 1 above)
Comment (samrg123) [reply src/vs/workbench/contrib/terminal/browser/links/terminalLinkManager.ts:233]: Good idea. I'll update the code

**Review by rheh — APPROVED** [review]
Perfect

**Comment by Tyriar** [comment]
Sorry for not getting to this, it looks like someone on the team got to it in the meantime in https://github.com/microsoft/vscode/issues/110509

**Pull request closed without merging**